File: ArraysAndHashing/ValidAnagram.py
class Solution(object):
    def isAnagram(self, s, t):
        if len(s) != len(t):
            return False
        sMap = {}
        tMap = {}
        for it in range(len(s)):
            sMap[s[it]] = sMap.get(s[it],0) + 1
            tMap[t[it]] = tMap.get(t[it],0) + 1
        return sMap == tMap
    
# Time complexity: O(n)
# Space complexity: O(n)

# Method 2: sort

class Solution(object):
    def isAnagram(self, s, t):
        return sorted(s) == sorted(t)

# Time complexity: O(nlogn)
# Space complexity: O(1)

File: ArraysAndHashing/ValidAnagram.cpp
#include <string>
#include <map>
#include <algorithm>

using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
    if(s.length() != t.length())
        return false;
    map<char,int> sMap;
    map<char,int> tMap;
    for(int it = 0 ; it < s.length() ; it++)
    {
        sMap[s[it]]++;
        tMap[t[it]]++;
    }
    if(sMap != tMap)
        return false;
    return true;
    }
};

//this solution has a time complexity of O(n) and space complexity of O(n)
//we can also sort the strings and compare them. The time complexity will be O(nlogn) and space complexity will be O(1)

class Solution {
public:
    bool isAnagram(string s, string t) {
    if(s.length() != t.length())
        return false;
    sort(s.begin(), s.end());
    sort(t.begin(), t.end());
    if(s != t)
        return false;
    return true;
    }
};

//this solution has a time complexity of O(nlogn) and space complexity of O(1)